enclave-runtime/src/dcap_attestation.rs

```diff
@@ -0,0 +1,340 @@
+	pub_k: &[u8; 32],
+	ocall_api: &A,
+	quoting_enclave_target_info: &sgx_target_info_t,
+	quote_size: u32,
+) -> SgxResult<Vec<u8>> {
+	// Generate app enclave report and include the enclave public key.
+	// The quote will be generated on top of this report and validate that the
+	// report as well as the public key inside it are coming from a legit
+	// intel sgx enclave.
+	let mut report_data: sgx_report_data_t = sgx_report_data_t::default();
+	report_data.d[..32].clone_from_slice(&pub_k[..]);
+
+	let app_report = match rsgx_create_report(quoting_enclave_target_info, &report_data) {
+		Ok(report) => {
+			debug!(
+				"rsgx_create_report creation successful. mr_signer: {:?}",
+				report.body.mr_signer.m
+			);
+			report
+		},
+		Err(e) => {
+			error!("rsgx_create_report creation failed. {:?}", e);
+			return Err(e)
+		},
+	};
+
+	// Retrieve quote from pccs for our app enclave.
+	debug!("Entering ocall_api.get_dcap_quote with quote size: {:?} ", quote_size);
+	let quote_vec = ocall_api.get_dcap_quote(app_report, quote_size)?;
+
+	// Check mrenclave of quote, to ensure the quote has not been tampered with
+	// while being on the untrusted side.
+	// This step is probably obsolete, as the QvE will check the quote as well on behalf
+	// of the target enclave.
+	let p_quote3: *const sgx_quote3_t = quote_vec.as_ptr() as *const sgx_quote3_t;
+	let quote3: sgx_quote3_t = unsafe { *p_quote3 };
+	if quote3.report_body.mr_enclave.m != app_report.body.mr_enclave.m {
+		error!("mr_enclave of quote and app_report are not matching");
+		error!("mr_enclave of quote: {:?}", quote3.report_body.mr_enclave.m);
+		error!("mr_enclave of quote: {:?}", app_report.body.mr_enclave.m);
+		return Err(sgx_status_t::SGX_ERROR_UNEXPECTED)
+	}
+
+	Ok(quote_vec)
+}
+
+pub fn generate_dcap_ecc_cert<A: EnclaveAttestationOCallApi>(
+	quoting_enclave_target_info: &sgx_target_info_t,
+	quote_size: u32,
+	ocall_api: &A,
+	skip_ra: bool,
+) -> EnclaveResult<(Vec<u8>, Vec<u8>)> {
+	let chain_signer = Ed25519Seal::unseal_from_static_file()?;
+	info!("[Enclave Attestation] Ed25519 signer pub key: {:?}", chain_signer.public().0);
+
+	let ecc_handle = SgxEccHandle::new();
+	let _result = ecc_handle.open();
+	let (prv_k, pub_k) = ecc_handle.create_key_pair()?;
+	info!("Enclave Attestation] Generated ephemeral ECDSA keypair:");
+	debug!(" pubkey X is {:02x}", pub_k.gx.iter().format(""));
+	debug!(" pubkey Y is {:02x}", pub_k.gy.iter().format(""));
+
+	let qe_quote = if !skip_ra {
+		let qe_quote = match retrieve_qe_dcap_quote(
+			&chain_signer.public().0,
+			ocall_api,
+			quoting_enclave_target_info,
+			quote_size,
+		) {
+			Ok(quote) => quote,
+			Err(e) => {
+				error!("[Enclave] Error in create_dcap_attestation_report: {:?}", e);
+				return Err(e.into())
+			},
+		};
+		qe_quote
+	} else {
+		Default::default()
+	};
+
+	// Verify the quote via qve enclave
+	let payload = ecdsa_quote_verification(qe_quote, ocall_api)?;
```

Comment:
Does this work if `skip_ra` is true? I would assume that it should fail

Reply:
Good point. Didn't think about that, thanks!

Reply:
Moved this call such that is not called when `skip_ra == true`. The implementation is now similar to `create_ra_report_and_signature(..)`.